### ISSUE TITLE
ANSI character codes displayed on prompt after piping output through less

### ISSUE THREAD

**Issue by drichardson** [microsoft/terminal#8525]
After piping `git log | less` and then pressing Q to quit, the PowerShell prompt displayed in Windows Terminal shows character codes, instead of being colored.

# Environment

```none
Windows build number: 10.0.19042.0
Windows Terminal version (if applicable): 1.4.3243.0

Any other software?
PowerShell Core: 7.1.0
```

# Steps to reproduce

1. Open PowerShell Core in Windows Termianl
1. Set the prompt to be colored with ANSI escape codes:
    ```
    function prompt { "`e[38;2;0;255;0mMyPrompt`e[0m>" }
    ```

1. Run git log | less. Using full paths here:
    ````
    & 'C:\Program Files\Git\bin\git.exe' log | & 'C:\Program Files\Git\usr\bin\less.exe'
    ````

1. Press Q to quit log.

# Expected behavior

Should see "MyPrompt>" in green.

# Actual behavior

Prompt looks like this:

    ←[38;2;0;255;0mMyPrompt←[0m>

# Workaround

The problem does not occur if Control+C is used to quit less.exe instead of Q.

The problem does not occur if `Out-Host -Paging` is used instead of less.exe.

    & 'C:\Program Files\Git\bin\git.exe' log | Out-Host -Paging

**Comment by DHowett** [comment]
`less` is disabling VT parsing. There is unfortunately nothing we can do about that: it is an explicit application request.

**Comment by drichardson** [comment]
The same problem occurs without less. For example, just using cat.exe (also from Git for Windows):

```
 echo "1`n2`n3`n" | & 'C:\Program Files\Git\usr\bin\cat.exe'
1
2
3

←[1m←[0m←[38;2;0;200;0mdoug←[0m←[38;2;200;100;100m←[0m>
```

I doubt cat does anything to the terminal regarding VT parsing.

**Comment by DHowett** [comment]
> I doubt

You would be surprised! I did an in-depth investigation here: https://github.com/microsoft/terminal/issues/4921#issuecomment-599135352

In short, the MSYS2 runtime shipped with Git does not manage the console's state appropriately. This applies equally to any MSYS2 clients.

**Comment by drichardson** [comment]
> In short, the MSYS2 runtime shipped with Git does not manage the console's state appropriately. This applies equally to any MSYS2 clients.

If that was true, I'd expect the same thing to happen when running one of the MSYS2 shells under Windows Terminal, but it does not.

I have MSYS2 installed (separately from Git for Windows) and can start it in Windows Terminal using this configuration entry:

```json
            {
                "guid": "{17da3cac-b318-431e-8a3e-7fcdefe6d114}",
                "name": "MINGW64 / MSYS2",
                "commandline": "C:/msys64/msys2_shell.cmd -defterm -here -no-start -mingw64",
                "startingDirectory": "C:/msys64/home/%USERNAME%",
                "icon": "C:/msys64/mingw64.ico",
            },
```

If I then do:

```
$ export PS1="\e[0;31mRED\e[m "
RED echo 1 |cat
1
```

or any of the other commands that screw up Windows Terminal when running PowerShell, it works just fine.

That also made me wonder about Windows Powershell (not PowerShell Core) and I tried the following:

```
PS C:\Users\doug> function prompt { $ESC = [char]27; "$ESC[38;2;0;255;0mMyPrompt$ESC[0m>" }
MyPrompt>echo "1`n2`n3`n" | & 'C:\Program Files\Git\usr\bin\cat.exe'
1
2
3

MyPrompt>
```

and it also worked just fine! (MyPrompt was green).

This leads me to believe that the problem is not with MSYS, but rather with Powershell Core. I'll open an issue up with them and reference this one, since it includes that interesting link @DHowett  had using Windows Event Recorder to observe `SetConsoleMode` calls.

**Comment by drichardson** [comment]
This apparently has been fixed in PowerShell, but I haven't got a chance to try a build with the fix yet to confirm.

**Comment by drichardson** [comment]
Confirmed fixed in PowerShell v7.2.0-preview.4. Perhaps this fix will address some of the other issues related to garbled console output.
